### DISCUSSION TITLE
Heroku Manual Deployment for VoIP Suite

### DISCUSSION THREAD

**Discussion by 0perationPrivacy** [0perationPrivacy/VoIP#133] [Announcements]
## Background

In the wake of GitHub and Heroku being "_hacked_" and their OAuth tokens being stolen, you can no longer connect your GitHub account to Heroku for automatic deployments as it authenticates using GitHub tokens. This should be temporary.

If you've already connected Heroku to GitHub before this incident, your GitHub account will remain connected but you still cannot use automatic Deployments. So when you update your forked repo, the Heroku app will not automatically pull the new code. Instead, the workaround is to manually push the code to the Heroku instance if you need to update your app.

## Updating Repo

When an update comes, you will follow the same procedure of first updating your fork by clicking on `Fetch Upstream` > `Fetch and Merge`

Here's where the new workaround starts:

---

## Manually Updating App

You need to manually push the updated code to your Heroku instance for the time being.

> Prerequisite: the git command must be installed in your machine. It is installed by default on Mac, Ubuntu and Kali Linux.

Take the following commands and copy it in your text editor for quick copy/pasting. You will copy/paste only 2 chunks of code and you should be done.

```
curl https://cli-assets.heroku.com/install.sh | sh
heroku login
git config --global user.email "me@me.com"

git clone https://github.com/0perationPrivacy/VoIP.git
cd VoIP
git commit --allow-empty -m "Empty test commit"
heroku git:remote -a MyVoIPSuite
git push heroku main --force

```
Download the txt file: [Heroku-Manual-Deploy.txt](https://github.com/0perationPrivacy/VoIP/files/8598476/Heroku-Manual-Deploy.txt)

From the above, change 2 things:
- `MyVoIPSuite` to the name of your actual application found in Heroku under https://dashboard.heroku.com/apps
- The GitHub repo link to your cloned repo. This is optional as the public repo is the source.

1. Copy the commands in a text editor and open up a terminal on an Linux VM.
1. Copy and paste the first chunk of commands
1. A browser window will open. Sign in to your Heroku account and authenticate with MFA.
  ![1-heroku-cli](https://user-images.githubusercontent.com/89225730/166136318-dba76a88-3fb9-4075-926c-f6fe1751eec3.png "Install Heroku CLI")
1. Copy/Paste the 2nd chunk of commands and wait for it to finish (10-15 sec)
  ![2-clone](https://user-images.githubusercontent.com/89225730/166136376-858b4d8e-63b9-4d5a-95fa-eff3de6e1b1e.png "Clone and Force Push")
1. Done. Your code should now be updated in Heroku. You can check under the `Activity` tab.

![3-deployed](https://user-images.githubusercontent.com/89225730/166136285-69b9a7ae-3e56-4622-8cda-24b5debd0ef9.png "Deployed")

The next time you want to update your code, you only have to:
1. Open the terminal and `cd` into the `VoIP` folder 
1. To update the local folder, run: `git pull`
1. To push the code back to Heroku, run: `git push Heroku main --force`

**Comment by buyer9** [comment]
**Here's how I worked this repair using Windows:**

First a Note: As a bit of a newbie to a lot of the services that are required to run this VoIPSuite, I had trouble figuring out what the error message that came from Heroku meant. But as it's explained nicely in this thread, it has to do with Heroku intentionally breaking "Automatic Deployments" from Github for security reasons. So, we now have to manually download the latest deployment from our Github account and then manually upload it to heroku. We will use GIT to handle the downloading, uploading and compiling. 



INSTALL GIT
First, You need to make sure some form of "GIT" is installed to your PC, as we will pull our updated deployments from github into GIT on our PC, and we will push the updated deployments out of our PC's GIT and into our accounts at Heroku. You can consider "GIT" to be kind of a specialized ftp program like Filezilla. 

As a PC user, I chose "Git for Windows" and chose the 64 bit standalone installer from here: https://git-scm.com/download/win

Before I ran the installer, I installed a 32-bit version of Notepad++: https://notepad-plus-plus.org/downloads/v8.3.3/   I did this because the GIT installer sort of asks for it.

Then I installed Git for Windows (Git-2.36.0-64-bit.exe) and accepted all the defaults for all the questions it asked me during the install. This worked out fine. (when it asked which text editor I wanted to use, I chose Notepad++) Notepad++ turned out to be unnecessary, but it's not harmful, and that's what I did. 

INSTALL HEROKU 
The instructions on the original post on this discussion page show the user how to download and install the Heroku CLI (Command Line Interface) in a linux environment. The code for that is:

curl https://cli-assets.heroku.com/install.sh | sh

But I'm on Windows. Fortunately, Heroku has a CLI help page for other platforms: https://devcenter.heroku.com/articles/heroku-cli 

There is a link "First-time GIT setup" (a prerequisite to installing heroku CLI)
It says to enter:

git config --global user.name "John Doe"
git config --global user.email johndoe@example.com

Since you have GIT installed, you are basically just writing these lines into the GIT config page on your PC. 

I opened up a command prompt with Administrative privileges. (as instructed)
It showed the following cursor: 
C:\Windows\system32

I then entered:

git config --global user.name "John Doe"
git config --global user.email me@me.com

(I use me@me.me because it appears later in the instructions. You can use it too.)

I later noticed that in the OP instructions to this thread, we actually set the global user email to me@me.com. So even though I set my config via the 2 lines above, I followed the rest of the instructions to the T. 

After making the tweaks to my GIT config, I downloaded and installed the 64 bit version of the Heroku CLI (heroku-x64.exe)> I got that file from the heroku help page referenced earlier:

[(https://devcenter.heroku.com/articles/heroku-cli)]

Now that the GIT and HEROKU CLIs are installed to your PC, you will issue commands via your Command Prompt terminal and these commands are understood and acted upon. 

GET YOUR FORK
Before you start downloading your fork of the VoIPSuite app, log into Github and make sure your branch is up-to-date. Go to your VoIPSuite github page and click on the "Fetch Upstream" link if needed. (https://github.com/yourUserGithubNameHere/VoIP)

You are looking for these magic words: "This branch is up to date with 0perationPrivacy/VoIP:main." Why? You will be downloading from your github account, so you need to make sure what's stored in your repository is the latest copy of VoIPSuite. 


HEROKU LOGIN
So now you can continue with the manual deployment instructions. Type into your command prompt: (and hit RETURN to execute the command)

heroku login

This causes the heroku website login page to open in a browser. Enter your heroku credentials. Get yourself logged in. 

next, go back to the command prompt and type the following (then hit RETURN or Enter)

git config --global user.email "me@me.com"

then type

git clone https://github.com/yourGithubUserNameHere/VoIP.git

obviously, replace yourGithubUserNameHere with your github username

for example, if I log into Github, and want to get to MY FORK of the VoIPSuite, my URL is: https://github.com/JoeUser/VoIP   JoeUser is my username

That git clone command will download a copy of your fork of VoIPSuite via GIT into your PC. 
In this example, it will get downloaded to C:\Windows\System32\VoIP

next type: 

cd VoIP

and hit return

that navigates us into the VoIP directory

next type:

git commit --allow-empty -m "Empty test commit"

and hit return (you are running a little connection test to GIT)

Next, you have to identify the receiving account on heroku for your coming upload

You'll have to personalize the following code "template" or example a bit before using it.

the code template is:

heroku git:remote -a tidbit

I will replace "tidbit" with a little bit of the URL of my VoIPSuite instance on Heroku. 

Specifically - if I want to use my VoIPSuite, I go to 
https://bada-BING.herokuapp.com/voip/dashboard  

So, bada-BING is the "tidbit" that I would put into the code template.

   Example - the template:
heroku git:remote -a tidbit
   becomes
heroku git:remote -a bada-BING

Just enter your app's name instead of tidbit or bada-BING and hit return

You have just set the receiving account on heroku for your coming upload

Next type

git push heroku main --force

and hit return. 

It will take a little time but this will upload and compile your app onto your heroku account. 

(Got an error? see below*)

VOILA, you have now made heroku happy and your VoIPSuite is back in action. 

Thanks to to OperationPrivacy for this swell app. I hope these instructions are helpful to somebody. 
================

*Got an error? such as: 
"fatal: destination path VoIP already exists and is not an empty directory."

You will get an error if this is not your first time doing this. That's because you have already creatged a VoIP folder inside C:\Windows\System32 and as a safety measure, GIT will not drop files in an existing GIT folder. Look into your PC. If you see see C:\Windows\System32\VoIP  it's because that folder was created when you 1st GIT-downloaded your VoIPSuite repository. 

My fixit was to do the following, and in fact, this would be the succinct code I would run each time I needed to download and push an update of VoIPSuite from github to heroku. 

open command prompt with administrator privileges
then type & hit enter
heroku login
then login to the heroku webpage that pops up
then go back to your command prompt and type:
git clone https://github.com/<yourGitHubUsername>/VoIP.git temp
and replace <yourGitHubUsername> with your github username. 

this puts all those repository files onto your PC here:
C:\Windows\System32\temp
next, cd into that new directory. Type
cd temp

now type
git commit --allow-empty -m "Empty test commit"
This tests git functionality. You will see some kind of successful response such as:
[main021a333] Empty test commit

next type:
heroku git:remote -a myCode123
  (Replacing myCode123 with the tidbit from the URL of my VoIPSuite instance on Heroku.)
finally, you can type
git push heroku main --force
that will compile and upload (force deploy) your updated github files to your herokuapp
You can close your command prompt when done.

You may have to close any opened VoIPsuite browser tabs and log back in.
